display_list/display_list_complexity_metal.cc

```diff
@@ -0,0 +1,758 @@
+DisplayListMetalComplexityCalculator::GetInstance() {
+  if (instance_ == nullptr) {
+    instance_ = new DisplayListMetalComplexityCalculator();
+  }
+  return instance_;
+}
+
+DisplayListMetalComplexityCalculator::MetalHelper::MetalHelper()
+    : is_complex_(false) {}
+
+unsigned int
+DisplayListMetalComplexityCalculator::MetalHelper::ComplexityScore() {
+  // We overflowed, return int_max
+  if (is_complex_) {
+    return std::numeric_limits<unsigned int>::max();
+  }
+
+  // Calculate the impact of saveLayer
+  unsigned int save_layer_complexity;
+  if (save_layer_count_ > 200) {
+    // m = 1/5
+    // c = 1
+    save_layer_complexity = (save_layer_count_ + 5) * 40000;
+  } else {
+    // m = 1/2
+    // c = 1
+    save_layer_complexity = (save_layer_count_ + 2) * 100000;
+  }
+
+  return complexity_score_ + save_layer_complexity;
+}
+
+void DisplayListMetalComplexityCalculator::MetalHelper::AccumulateComplexity(
+    unsigned int complexity) {
+  // Check to see if we will overflow by accumulating this complexity score
+  if (std::numeric_limits<unsigned int>::max() - complexity < complexity) {
+    is_complex_ = true;
+    return;
+  }
+
+  complexity_score_ += complexity;
+}
+
+void DisplayListMetalComplexityCalculator::MetalHelper::setAntiAlias(bool aa) {
+  current_paint_.setAntiAlias(aa);
+}
+
+void DisplayListMetalComplexityCalculator::MetalHelper::setStyle(
+    SkPaint::Style style) {
+  current_paint_.setStyle(style);
+}
+
+void DisplayListMetalComplexityCalculator::MetalHelper::setStrokeWidth(
+    SkScalar width) {
+  current_paint_.setStrokeWidth(width);
+}
+
+void DisplayListMetalComplexityCalculator::MetalHelper::saveLayer(
+    const SkRect* bounds,
+    const SaveLayerOptions options) {
+  if (is_complex_) {
+    return;
+  }
+  // saveLayer seems to have two trends; if the count is < 200,
+  // then the individual cost of a saveLayer is higher than if
+  // the count is > 200.
+  save_layer_count_++;
+}
+
+void DisplayListMetalComplexityCalculator::MetalHelper::drawColor(
+    SkColor color,
+    SkBlendMode mode) {
+  if (is_complex_) {
+    return;
+  }
+  // Placeholder value here. This is a relatively cheap operation.
+  AccumulateComplexity(50);
+}
+
+void DisplayListMetalComplexityCalculator::MetalHelper::drawPaint() {
+  if (is_complex_) {
+    return;
+  }
+  // Placeholder value here. This can be cheap (e.g. effectively a drawColor),
+  // or expensive (e.g. a bitmap shader with an image filter)
+  AccumulateComplexity(50);
+}
+
+// Normalise DrawLine to have a score of 100 when drawn with approximate line
+// length of 100, AA disabled and a hairline stroke
+// This represents a wall clock time of around 0.0005 milliseconds
+void DisplayListMetalComplexityCalculator::MetalHelper::drawLine(
+    const SkPoint& p0,
+    const SkPoint& p1) {
+  if (is_complex_) {
+    return;
+  }
+  // The performance penalties seem fairly consistent percentage-wise
+  float non_hairline_penalty = 1.0f;
```

Comment:
Right now this suggestion is a premature optimization, but just want to throw it out there in case it's useful later:

Consider avoiding floating point. You can do that by multiplying these factors by 10, and then dividing by 10 later when you need the final result.

Other strategies might be needed below, but we can delay thinking about them until they're needed.